File: quant1x/level1/security_bars.rs
// K线类型 (mimicking C++ level1::KLineType)
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum KLineType {
    _5Min = 0,    // 5分钟K线
    _15Min = 1,   // 15分钟K线
    _30Min = 2,   // 30分钟K线
    _1Hour = 3,   // 1小时K线
    Daily = 4,    // 日K线
    Weekly = 5,   // 周K线
    Monthly = 6,  // 月K线
    Exhq1Min = 7, // 扩展市场1分钟
    _1Min = 8,    // 普通1分钟K线
    RiK = 9,      // 日K线(同DAILY)
    _3Month = 10, // 季K线
    Yearly = 11,  // 年K线
}

use super::sequence_id;
use crate::level1::commands::*;
use crate::level1::protocol::{Request, RequestHeader, Response, ResponseHeader};
use crate::std::BinaryStream;

#[derive(Debug, Clone, Default)]
struct SecurityBarsParameter {
    market: u16,
    code: [u8; 6],
    category: u16,
    i_field: u16,
    start: u16,
    count: u16,
}

#[allow(dead_code)]
#[derive(Debug, Clone)]
pub struct SecurityBarsRequest {
    header: RequestHeader,
    param: SecurityBarsParameter,
    padding: [u8; 10],
    is_index: bool,
}

#[allow(dead_code)]
impl SecurityBarsRequest {
    pub fn new(security_code: &str, category: u16, start: u16, count: u16) -> Self {
        Self::with_frequency(security_code, category, start, count, 1)
    }

    pub fn with_frequency(
        security_code: &str,
        category: u16,
        start: u16,
        count: u16,
        frequency: u16,
    ) -> Self {
        let mut header = RequestHeader::new();
        header.zip_flag = crate::level1::protocol::zlib_flag::UNCOMPRESSED;
        header.seq_id = sequence_id();
        header.packet_type = 0x00;
        header.method = SECURITY_BARS;

        let (market_id, _flag, pure) = crate::exchange::detect_market(security_code);
        let mut code_bytes = [0u8; 6];
        let sym = pure.as_bytes();
        let copy_len = std::cmp::min(sym.len(), code_bytes.len());
        code_bytes[..copy_len].copy_from_slice(&sym[..copy_len]);

        let is_index = crate::exchange::assert_index_by_market_and_code(market_id, pure.as_str());

        SecurityBarsRequest {
            header,
            param: SecurityBarsParameter {
                market: market_id as u16,
                code: code_bytes,
                category,
                i_field: frequency,
                start,
                count,
            },
            padding: [0u8; 10],
            is_index,
        }
    }

    pub fn is_index(&self) -> bool {
        self.is_index
    }

    fn code_string(&self) -> String {
        let nul = self
            .param
            .code
            .iter()
            .position(|&b| b == 0)
            .unwrap_or(self.param.code.len());
        String::from_utf8_lossy(&self.param.code[..nul]).into_owned()
    }
}

impl Request for SecurityBarsRequest {
    fn header(&self) -> &RequestHeader {
        &self.header
    }

    fn header_mut(&mut self) -> &mut RequestHeader {
        &mut self.header
    }

    fn serialize_payload(&mut self) -> Vec<u8> {
        let mut payload = BinaryStream::new();
        payload.push_u16(self.param.market);
        payload.push_byte_array(&self.param.code);
        payload.push_u16(self.param.category);
        payload.push_u16(self.param.i_field);
        payload.push_u16(self.param.start);
        payload.push_u16(self.param.count);
        payload.push_byte_array(&self.padding);
        payload.data().clone()
    }

    fn payload_string(&self) -> String {
        format!(
            "{{Market:{}, Code:{}, Category:{}, I:{}, Start:{}, Count:{}, padding:{}}}",
            self.param.market,
            self.code_string(),
            self.param.category,
            self.param.i_field,
            self.param.start,
            self.param.count,
            hex::encode(self.padding)
        )
    }
}

#[derive(Debug, Clone)]
pub struct SecurityBar {
    pub open: f64,
    pub close: f64,
    pub high: f64,
    pub low: f64,
    pub vol: f64,
    pub amount: f64,
    pub year: i32,
    pub month: i32,
    pub day: i32,
    pub hour: i32,
    pub minute: i32,
    pub datetime: String,
    pub up_count: u16,
    pub down_count: u16,
}

impl SecurityBar {
    pub fn new() -> Self {
        Self {
            open: 0.0,
            close: 0.0,
            high: 0.0,
            low: 0.0,
            vol: 0.0,
            amount: 0.0,
            year: 0,
            month: 0,
            day: 0,
            hour: 0,
            minute: 0,
            datetime: String::new(),
            up_count: 0,
            down_count: 0,
        }
    }
}

#[allow(dead_code)]
#[derive(Debug, Clone)]
pub struct SecurityBarsResponse {
    header: ResponseHeader,
    pub count: u16,
    pub list: Vec<SecurityBar>,
    // additional metadata typically provided by the request/response constructor
    pub is_index: bool,
    pub category: u16,
}

#[allow(dead_code)]
impl SecurityBarsResponse {
    pub fn new() -> Self {
        Self {
            header: ResponseHeader::new(),
            count: 0,
            list: Vec::new(),
            is_index: false,
            category: 0,
        }
    }
    pub fn new_with(is_index: bool, category: u16) -> Self {
        Self {
            header: ResponseHeader::new(),
            count: 0,
            list: Vec::new(),
            is_index,
            category,
        }
    }
}

impl Response for SecurityBarsResponse {
    fn header(&self) -> &ResponseHeader {
        &self.header
    }

    fn header_mut(&mut self) -> &mut ResponseHeader {
        &mut self.header
    }

    fn deserialize_body(&mut self, data: &[u8]) -> Result<(), crate::std::DeserializeError> {
        let mut bs = BinaryStream::from_vec(data.to_vec());
        self.count = bs.get_u16()?;
        self.list.clear();
        self.list.reserve(self.count as usize);

        let mut pre_diff_base: i64 = 0;
        for _ in 0..self.count {
            let mut e = SecurityBar::new();

            if (self.category as i32) < 4 || self.category == 7 || self.category == 8 {
                let zipday = bs.get_u16()? as u32;
                let tminutes = bs.get_u16()?;
                let (y, m, d, hh, mm) =
                    super::get_datetime_from_u32(self.category as i32, zipday, tminutes);
                e.year = y;
                e.month = m;
                e.day = d;
                e.hour = hh;
                e.minute = mm;
            } else {
                let zipday = bs.get_u32()?;
                let (y, m, d, hh, mm) =
                    super::get_datetime_from_u32(self.category as i32, zipday, 0);
                e.year = y;
                e.month = m;
                e.day = d;
                e.hour = hh;
                e.minute = mm;
            }
            e.datetime = format!(
                "{:04}-{:02}-{:02} {:02}:{:02}:00",
                e.year, e.month, e.day, e.hour, e.minute
            );

            let mut price_open_diff = bs.varint_decode()?;
            let price_close_diff = bs.varint_decode()?;
            let price_high_diff = bs.varint_decode()?;
            let price_low_diff = bs.varint_decode()?;

            let ivol = bs.get_u32()?;
            e.vol = super::int_to_float64(ivol);

            let dbvol = bs.get_u32()?;
            e.amount = super::int_to_float64(dbvol);

            e.open = (price_open_diff + pre_diff_base) as f64 / 1000.0;
            price_open_diff += pre_diff_base;

            e.close = (price_open_diff + price_close_diff) as f64 / 1000.0;
            e.high = (price_open_diff + price_high_diff) as f64 / 1000.0;
            e.low = (price_open_diff + price_low_diff) as f64 / 1000.0;

            pre_diff_base = price_open_diff + price_close_diff;

            if self.is_index {
                e.up_count = bs.get_u16()?;
                e.down_count = bs.get_u16()?;
            }

            self.list.push(e);
        }
        Ok(())
    }

    fn body_string(&self) -> String {
        format!(
            "{{Count:{}, Parsed:{}, IsIndex:{}}}",
            self.count,
            self.list.len(),
            self.is_index
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deserialize_sample_matches_cpp_output() {
        let hex_data = "05002bff3401a52910134982d4834e07eb2f4f2eff340102060e4a8a70db4dca40934e2fff3401440a0f4aef5a734e3b6c234f30ff340141191f515cd8094f6d64ba4f31ff34014d102c4398098b4e44b03c4f";
        let buf = hex::decode(hex_data).unwrap();
        let mut resp = SecurityBarsResponse::new_with(false, 9);
        let _ = resp.deserialize(&buf);
        assert_eq!(resp.count as usize, resp.list.len());
        assert!(resp.list.len() > 0 || resp.count == 0);
    }
}

File: quant1x/level1/security_bars.py
# -*- coding: UTF-8 -*-
import struct
from enum import Enum
from typing import List, Tuple
from dataclasses import dataclass

from quant1x.level1 import protocol, helpers
from quant1x.exchange import code as exchange_code

SECURITY_BARS_MAX = 800

class KLineType(Enum):
    _5MIN = 0
    _15MIN = 1
    _30MIN = 2
    _1HOUR = 3
    DAILY = 4
    WEEKLY = 5
    MONTHLY = 6
    EXHQ_1MIN = 7
    _1MIN = 8
    RI_K = 9
    _3MONTH = 10
    YEARLY = 11

    @staticmethod
    def to_string(ktype: 'KLineType') -> str:
        return ktype.name

@dataclass
class SecurityBar:
    Open: float = 0.0
    Close: float = 0.0
    High: float = 0.0
    Low: float = 0.0
    Vol: float = 0.0
    Amount: float = 0.0
    Year: int = 0
    Month: int = 0
    Day: int = 0
    Hour: int = 0
    Minute: int = 0
    DateTime: str = ""
    UpCount: int = 0
    DownCount: int = 0

    def __str__(self):
        return (f"Open: {self.Open} Close: {self.Close} High: {self.High} Low: {self.Low} "
                f"Vol: {self.Vol} Amount: {self.Amount} Year: {self.Year} Month: {self.Month} "
                f"Day: {self.Day} Hour: {self.Hour} Minute: {self.Minute} DateTime: {self.DateTime} "
                f"UpCount: {self.UpCount} DownCount: {self.DownCount}")

class SecurityBarsRequest:
    def __init__(self, security_code: str, category: int, start: int, count: int):
        self.zip_flag = protocol.FLAG_UNCOMPRESSED
        self.seq_id = protocol.sequence_id()
        self.packet_type = 0x00
        self.method = protocol.COMMAND_SECURITY_BARS
        
        self.category = category
        self.i = 1
        self.start = start
        self.count = count
        
        market_id, _, symbol = exchange_code.detect_market(security_code)
        self.market = market_id.value if hasattr(market_id, 'value') else market_id
        self.code = symbol
        self.is_index = exchange_code.assert_index_by_market_and_code(market_id, symbol)
        
        self.padding = bytes.fromhex("00000000000000000000")

    def serialize(self) -> bytes:
        # Body: Market(2) + Code(6) + Category(2) + I(2) + Start(2) + Count(2) + Padding(10)
        # Total Body = 26 bytes
        body_len = 2 + 6 + 2 + 2 + 2 + 2 + len(self.padding)
        pkg_len = body_len + 2
        
        header = struct.pack('<B I B H H H', 
                             self.zip_flag, self.seq_id, self.packet_type, 
                             pkg_len, pkg_len, self.method)
        
        code_bytes = self.code.encode('ascii')
        if len(code_bytes) < 6:
            code_bytes = code_bytes + b'\x00' * (6 - len(code_bytes))
        else:
            code_bytes = code_bytes[:6]
            
        body = struct.pack('<H 6s H H H H', 
                           self.market, code_bytes, self.category, self.i, self.start, self.count)
        return header + body + self.padding

class SecurityBarsResponse:
    def __init__(self, is_index: bool, category: int):
        self.count = 0
        self.list: List[SecurityBar] = []
        self.is_index = is_index
        self.category = category

    def deserialize(self, data: bytes):
        if len(data) < 2:
            return
            
        self.count = struct.unpack('<H', data[:2])[0]
        pos = 2
        
        pre_diff_base = 0
        
        for _ in range(self.count):
            if pos >= len(data):
                break
                
            e = SecurityBar()
            year = 0
            month = 0
            day = 0
            hour = 15
            minute = 0
            
            if self.category < 4 or self.category == 7 or self.category == 8:
                if pos + 4 > len(data): break
                zipday = struct.unpack('<H', data[pos:pos+2])[0]
                pos += 2
                tminutes = struct.unpack('<H', data[pos:pos+2])[0]
                pos += 2
                
                year, month, day, hour, minute = helpers.get_datetime_from_uint32(self.category, zipday, tminutes)
            else:
                if pos + 4 > len(data): break
                zipday = struct.unpack('<I', data[pos:pos+4])[0]
                pos += 4
                year = int(zipday / 10000)
                month = int((zipday % 10000) / 100)
                day = int(zipday % 100)
                
            e.Year = year
            e.Month = month
            e.Day = day
            e.Hour = hour
            e.Minute = minute
            e.DateTime = f"{year:04d}-{month:02d}-{day:02d} {hour:02d}:{minute:02d}:00"
            
            price_open_diff, pos = helpers.varint_decode(data, pos)
            price_close_diff, pos = helpers.varint_decode(data, pos)
            price_high_diff, pos = helpers.varint_decode(data, pos)
            price_low_diff, pos = helpers.varint_decode(data, pos)
            
            if pos + 8 > len(data): break
            ivol = struct.unpack('<I', data[pos:pos+4])[0]
            pos += 4
            e.Vol = helpers.int_to_float64(ivol)
            
            dbvol = struct.unpack('<I', data[pos:pos+4])[0]
            pos += 4
            e.Amount = helpers.int_to_float64(dbvol)
            
            e.Open = float(price_open_diff + pre_diff_base) / 1000.0
            price_open_diff += pre_diff_base
            
            e.Close = float(price_open_diff + price_close_diff) / 1000.0
            e.High = float(price_open_diff + price_high_diff) / 1000.0
            e.Low = float(price_open_diff + price_low_diff) / 1000.0
            
            pre_diff_base = price_open_diff + price_close_diff
            
            if self.is_index:
                if pos + 4 > len(data): break
                e.UpCount = struct.unpack('<H', data[pos:pos+2])[0]
                pos += 2
                e.DownCount = struct.unpack('<H', data[pos:pos+2])[0]
                pos += 2
                
            self.list.append(e)
